### PR TITLE
3.x

### PR DESCRIPTION
**Pull request by tomkis1**
You can try it via npm `npm install redux-elm@3.x-latest`

- [x] Unit tests
- [x] Update docs
- [x] Code comments

### CODE REVIEW

File: docs/custom-matchers/writing-custom-matcher.md [hunk 1]
```diff
@@ -1,9 +1,10 @@
 ## Writing Custom Matcher
 
-Matcher is a function that takes `pattern` as an argument and returns a function that takes an Action as its argument. The function returns either `false` when the Action does not match or, when it does, an `Object` consisting from three fields `unwrap`, `wrap`, and `args`.
+Matcher is a function that takes `pattern` as an argument and returns a function that takes an Action as its argument. The function returns either `false` when the Action does not match or, when it does, an `Object` consisting from four fields `id`, `unwrappedType`, `wrap`, and `args`.
 
-* `unwrap` is unwrapped action type
-* `wrap` is stripped part of the action which can be used for wrapping the action
+* `id` is unique string identifying matched action, this is in 99% same as provided pattern
```
Comment (matthewgertner) [inline docs/custom-matchers/writing-custom-matcher.md]: Not sure what this means

File: docs/custom-matchers/writing-custom-matcher.md [hunk 2]
```diff
@@ -15,8 +16,9 @@ const endsWithMatcher = pattern => {
       const wrapRegExp = new RegExp(`(.*)\.${pattern}$`); // In production application pattern should definitely be escaped!
 
       return {
-        unwrap: pattern,
-        wrap: action.type.match(wrapRegExp)[1],
+        id: `ends-with${pattern}`,
```
Comment (matthewgertner) [inline docs/custom-matchers/writing-custom-matcher.md]: Not `ends-with-${pattern}`?

File: docs/custom-matchers/writing-custom-matcher.md [hunk 3]
```diff
@@ -15,8 +16,9 @@ const endsWithMatcher = pattern => {
...
+        unwrappedType: pattern,
+        wrap: type => `${action.type.match(wrapRegExp)[1]}.${type}`,
```
Comment (matthewgertner) [inline docs/custom-matchers/writing-custom-matcher.md]: Maybe it's because I'm reading this out of context, but I have no idea what this is or why we need it.

File: src/SagaRepository.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+  constructor() {
+    this.sagas = {};
+  }
+
+  /**
+   * Mounts Saga instance
+   *
+   * @param {SagaAbstraction} Saga abstraction type (rxjs/redux-saga)
+   * @param {Srting} ID of the Saga
```
Comment (matthewgertner) [inline src/SagaRepository.js]: Typo

File: src/SagaRepository.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  ) {
+    if (!this.sagas[sagaId]) {
+      const sagaInstance = new SagaAbstraction(saga, model);
+      sagaInstance.subscribe(dispatch);
+
+      this.sagas[sagaId] = sagaInstance;
+    } else {
+      warn(
+        'The Saga instance has already been mounted, this basically mean ' +
+        'that your Updaters do not form a tree hierarchy, please be sure that ' +
```
Comment (matthewgertner) [inline src/SagaRepository.js]: "tree hierarchy" is kind of redundant

File: src/SagaRepository.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        'any used Updater is wrapped by any other Updater (except root updater). ' +
```
Comment (matthewgertner) [inline src/SagaRepository.js]: I would say "every Updater is wrapped by another Updater..."

File: src/Updater.js [hunk 7]
```diff
@@ -1,154 +1,117 @@
...
+              );
             }
           });
         }
 
-        return reduction;
+        return mutatedModel;
       } else {
-        return model;
+        return reducer(model, action);
```
Comment (matthewgertner) [inline src/Updater.js]: Why did you change this?

File: src/forwardTo.js [hunk 8]
```diff
@@ -1,19 +1,16 @@
...
+ * @return {Function} Dispatch which automatically wraps all the actions
  */
 export default (dispatch, ...types) => {
   if (types.length === 0) {
     return dispatch;
   } else {
-    if (types.some(type => ~type.toString().indexOf('.'))) {
-      throw new Error('Action type can\'t contain a dot');
-    }
-
```
Comment (matthewgertner) [inline src/forwardTo.js]: Why did you remove this? You really need to annotate your PRs.

File: src/matching/MatchingReducerFactory.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
+import { matcher as defaultMatcher } from './matchers';
+import { warn } from '../utils/logger';
+
+const identity = value => value;
+
+/**
+ * @class MatchingReducerFactory
+ *
+ * A simple pattern matching abstraction
+ * for redux reducers.
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: "Redux"

File: src/matching/MatchingReducerFactory.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  constructor(initialAppState, defaultMatcherImpl = defaultMatcher) {
+    this.defaultMatcherImpl = defaultMatcherImpl;
+    this.initialAppState = initialAppState;
+    this.matchersWithHandlers = [];
+  }
+
+  /**
+   * Registers action handler
+   *
+   * @param {String} String pattenr
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: Typo

File: src/matching/MatchingReducerFactory.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      actionHandler
+    });
+
+    return this;
+  }
+
+  /**
+   * Creates the Reducer
+   *
+   * @return {Function} plain old redux Reducer
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: "Redux" should be capitalized (but "reducer" really shouldn't)

File: src/matching/MatchingReducerFactory.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+   */
+  toReducer() {
+    return (appState = this.initialAppState, action) => {
+      if (action) {
+        if (!action.type) {
+          // Logging is technically impurity yet this serves
+          // just for developer to eliminate any potential issues
+          // in development phase.
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: Can't you detect production mode like Redux does?

File: src/matching/MatchingReducerFactory.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+          warn('It looks like you have provided invalid action. ' +
+               'Please make sure that you have dispatched an object with at least ' +
+               'one field (type) identifiying type of the dispatched action. Or ' +
+               'make sure that you are using appropriate middleware (eg. redux-thunk) ' +
+               'which translates non-standard actions to plain old Redux actions in ' +
+               'form of objects.');
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: "It looks like you have provided an invalid action. Make sure that the dispatched action has a `type` property or use the appropriate middleware (e.g. redux-thunk) to translate non-standard actions into plain old Redux action objects."

File: src/matching/MatchingReducerFactory.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+          }, actionHandler }) =>
+            actionHandler(
+              partialReduction,
+              {
+                ...action,
+                type: unwrappedType,
+                matching: {
+                  id: `${parentId}${id}`,
+                  wrap: type => parentWrap(wrap(type)),
+                  args: Object.assign(parentArgs, args)
```
Comment (matthewgertner) [inline src/matching/MatchingReducerFactory.js]: You really want to mutate `parentArgs`?

File: src/matching/matchers/matcher.js [hunk 15]
```diff
@@ -11,17 +11,21 @@ export default pattern => {
   return action => {
     if (action.type === pattern) {
       return {
-        wrap: '',
-        unwrap: action.type,
+        id: pattern,
```
Comment (matthewgertner) [inline src/matching/matchers/matcher.js]: Annotation would be great!

File: src/sagas/ReduxSaga.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+   *
+   * @param {Generator} Saga implementation
+   * @param {any} initial model
+   */
+  constructor(saga, model) {
+    this.updateModel(model);
+
+    this.dispatchSubject = new Subject();
+    this.subscribeSubject = new Subject();
+    this.subscribtion = null;
```
Comment (matthewgertner) [inline src/sagas/ReduxSaga.js]: `subscription`

File: src/sagas/RxjsSaga.js [hunk 17]
```diff
@@ -0,0 +1,69 @@
+import { Subject } from 'rxjs';
+
+const isAction = any => any && typeof any === 'object' && !!any.type;
+
+/**
+ * @class ReduxSaga
+ *
+ * RxJS implementation of Saga Abstraction.
+ */
+export default class RxjsSaga {
```
Comment (matthewgertner) [inline src/sagas/RxjsSaga.js]: The name looks a bit weird. Maybe you should just call it `RxSaga`.

File: src/sagas/RxjsSaga.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+   *
+   * @param {Function} Saga implementation
+   * @param {any} initial model
+   */
+  constructor(saga, model) {
+    this.updateModel(model);
+    this.subject = new Subject();
+    this.saga$ = saga(this.subject);
+
+    this.subscribtion = null;
```
Comment (matthewgertner) [inline src/sagas/RxjsSaga.js]: `subscription`

File: src/storeEnhancer.js [hunk 19]
```diff
@@ -10,20 +13,29 @@ export default createStore => (reducer, initialAppState) => {
...
+      // "almost" pure
       setTimeout(() => {
         if (wrappedDispatch) {
-          callback(wrappedDispatch);
+          effect(wrappedDispatch);
         } else {
-          console.warn(
+          warn(
             'There\'s been attempt to execute effects ' +
             'yet proper creating of store has not been finished yet'
```
Comment (matthewgertner) [inline src/storeEnhancer.js]: "Attempt to execute effects before store creation has completed"

File: src/storeEnhancer.js [hunk 20]
```diff
@@ -32,18 +44,31 @@ export default createStore => (reducer, initialAppState) => {
...
   });
 
-  wrappedDispatch = (...args) => callWithEffects(() => store.dispatch(...args));
+  // Wrapped dispatch executes all the effects
+  wrappedDispatch = (...args) => callFnWithEffects(() => store.dispatch(...args));
 
   return {
     ...store,
     dispatch: wrappedDispatch,
+    // We must no forgot about mandatory sagaRepository & effectExecutor
```
Comment (matthewgertner) [inline src/storeEnhancer.js]: "not"

File: src/utils/logger.js [hunk 21]
```diff
@@ -0,0 +1,7 @@
+const log = (message, level) => {
+  if (typeof console !== 'undefined' && typeof console[level] === 'function') {
+    console[level](`WARNING: ${message}`);
+  }
+};
+
+export const warn = message => log(message, 'error');
```
Comment (matthewgertner) [inline src/utils/logger.js]: Why 'error' rather than 'warn'?

File: src/view.js [hunk 22]
```diff
@@ -61,6 +61,6 @@ export default View => class ReduxElmView extends Component {
   }
 
   render() {
-    return createElement(View, { ...this.props, dispatch: this.dispatch });
+    return createEagerElement(View, { ...this.props, dispatch: this.dispatch });
```
Comment (matthewgertner) [inline src/view.js]: Annotation!

**Comment by matthewgertner** [comment]
I didn't review the test stuff but can have a look if you think it makes sense.

File: docs/custom-matchers/writing-custom-matcher.md (hunk 2 above)
Comment (tomkis1) [inline docs/custom-matchers/writing-custom-matcher.md]: Technically it does not matter, because this ID just needs to be unique, so unless somebody writes custom matcher which generates the same ID everything's fine.

Comment (tomkis1) [inline docs/custom-matchers/writing-custom-matcher.md]: Which reminds me that we should probably prefix all the built-in matchers with something like `@@redux-saga`

File: docs/custom-matchers/writing-custom-matcher.md (hunk 3 above)
Comment (tomkis1) [inline docs/custom-matchers/writing-custom-matcher.md]: https://github.com/salsita/redux-elm/blob/54630bdbf0638c39af66fe3092432c298d664995/docs/custom-matchers/writing-custom-matcher.md Should be explained here:

> `wrap` is a function which can be used for wrapping the action back

File: src/Updater.js (hunk 7 above)
Comment (tomkis1) [inline src/Updater.js]: It definitely must have been wrong, because when there's no action provided it should call the underlying reducer to obtain empty model.

File: src/forwardTo.js (hunk 8 above)
Comment (tomkis1) [inline src/forwardTo.js]: It's been moved to `wrapAction` instead, because technically it has nothing to do with forwarding.

File: src/matching/MatchingReducerFactory.js (hunk 12 above)
Comment (tomkis1) [inline src/matching/MatchingReducerFactory.js]: I don't think it's desired here. Because this potentially leads to runtime error, so I don't think that "catching" this error and logging to console even in production would be anyhow dangerous.

File: src/matching/MatchingReducerFactory.js (hunk 14 above)
Comment (tomkis1) [inline src/matching/MatchingReducerFactory.js]: No :-)

File: src/utils/logger.js (hunk 21 above)
Comment (tomkis1) [inline src/utils/logger.js]: It's better to show it in console as error.

File: src/view.js (hunk 22 above)
Comment (tomkis1) [inline src/view.js]: The name of the function in the library has changed... no functional changes though.

File: src/Updater.js (hunk 7 above)
Comment (matthewgertner) [inline src/Updater.js]: IMO would be clearer to say `reducer(model, null)`.
